Tribler/community/multichain/community.py

```diff
@@ -0,0 +1,334 @@
+        :param message:
+        """
+        block = DatabaseBlock.from_signature_response_message(message)
+        self._logger.info("Persisting sr: %s" % base64.encodestring(block.id))
+        self.persistence.add_block(block)
+
+    def publish_request_block_message(self, candidate, sequence_number=-1):
+        """
+        Request a specific block from a chain of another candidate.
+        :param candidate: The candidate that the block is requested from
+        :param sequence_number: The requested sequence_number or default the latest sequence number
+        """
+        self._logger.info("Crawler: Requesting Block:%s" % sequence_number)
+        meta = self.get_meta_message(BLOCK_REQUEST)
+
+        message = meta.impl(authentication=(self.my_member,),
+                            distribution=(self.claim_global_time(),),
+                            destination=(candidate,),
+                            payload=(sequence_number,))
+        self.dispersy.store_update_forward([message], False, False, True)
+
+    def received_request_block(self, messages):
+        for message in messages:
+            requested_sequence_number = message.payload.requested_sequence_number
+            self._logger.info("Crawler: Received request for block: %s" % requested_sequence_number)
+            self.publish_block(message.candidate, requested_sequence_number)
+
+    def publish_block(self, candidate, sequence_number):
+        requested_block = self.persistence.get_by_sequence_number_public_key(sequence_number, self._public_key)
+        if requested_block:
+            self._logger.info("Crawler: Sending block: %s" % sequence_number)
+            meta = self.get_meta_message(BLOCK_RESPONSE)
+
+            message = meta.impl(authentication=(self.my_member,),
+                                distribution=(self.claim_global_time(),),
+                                destination=(candidate,),
+                                payload=requested_block.to_payload())
+
+            self.dispersy.store_update_forward([message], False, False, True)
+        else:
+            self._logger.info("Crawler: Received invalid request for block: %s" % sequence_number)
+
+    def received_block_response(self, messages):
+        """
+        We've received a valid block response and must process this message.
+        """
+        self._logger.info("Crawler: Valid %s block response(s) received." % len(messages))
+        for message in messages:
+            block = DatabaseBlock.from_block_response_message(message)
+            # Create the hash of the message
+            if not self.persistence.contains(block.id):
+                self._logger.info("Crawler: Persisting sr: %s" % base64.encodestring(block.id))
+                try:
+                    self.persistence.add_block(block)
+                except IntegrityError:
+                    # A race condition can occur between here and
```

Comment:
Can't happen. Dispersy is single threaded. If a block wasn't stored in the database at line 273, it can't have been stored now.